### PR TITLE
New initialisation method when passin a `df.Field` using the new xarray functionality.

### PR DESCRIPTION
**Pull request by lang-m**
Test
- field with `mesh.n = (100, 100, 10)`
- creation of a new field with `mesh.n = (10, 10, 10)` and passing the old field to `value`

Performance improvement:
- old implementation (using that the field is callable): ~ 6.5 s
- new implementation (interpolation done by `xarray`): ~ 5 ms

@swapneelap This is probably the first use case for your new method.

### CODE REVIEW

**Comment by swapneelap** [comment]
@lang-m Happy to see that it is useful to the project! :)
Great performance improvement BTW.

File: discretisedfield/field.py [hunk 1]
```diff
@@ -3794,3 +3795,88 @@ def from_xarray(cls, xa):
...
+
+
+@_as_array.register(collections.abc.Callable)
+def _(val, mesh, dim, dtype):
+    # will only be called on user input
+    # dtype must be specified by the user for complex values
+    array = np.empty((*mesh.n, dim), dtype=dtype)
+    for index, point in zip(mesh.indices, mesh):
+        res = val(point)
+        array[index] = res
```
Comment (marijanbeg) [inline discretisedfield/field.py:3829]: `array[index] = val(point)`

File: discretisedfield/field.py [hunk 2]
```diff
@@ -3794,3 +3795,88 @@ def from_xarray(cls, xa):
...
+@_as_array.register(Field)
+def _(val, mesh, dim, dtype):
+    if mesh.region not in val.mesh.region:
+        raise ValueError(f'The region {val.mesh.region} of the provided field'
+                         f' does not contain the region {mesh.region} of the'
+                         ' field that is being created.')
+    value = val.to_xarray().sel(x=mesh.midpoints.x,
+                                y=mesh.midpoints.y,
+                                z=mesh.midpoints.z,
+                                method='nearest').data
```
Comment (marijanbeg) [inline discretisedfield/field.py:3842]: Beautiful!

File: discretisedfield/field.py [hunk 3]
```diff
@@ -3794,3 +3795,88 @@ def from_xarray(cls, xa):
...
+        return value.reshape(mesh.n + (-1,))
+    return value
+
+
+@_as_array.register(dict)
+def _(val, mesh, dim, dtype):
+    # will only be called on user input
+    # dtype must be specified by the user for complex values
+    if dtype is None:
+        dtype = np.float64
```
Comment (marijanbeg) [inline discretisedfield/field.py:3854]: `dtype = dtype or np.float64` (short-circuiting)

File: discretisedfield/field.py [hunk 4]
```diff
@@ -3794,3 +3795,88 @@ def from_xarray(cls, xa):
...
+    if 'default' in val and not callable(val['default']):
+        fill_value = val['default']
+    else:
+        fill_value = np.nan
```
Comment (marijanbeg) [inline discretisedfield/field.py:3858]: single-line?

**Review by marijanbeg — COMMENTED** [review]
Looks very good! Just added a few minor suggestions.

Is there a reason why we want all `_as_array` in `Field` class and not in `dfu`?

**Comment by lang-m** [comment]
@marijanbeg Could you please have a look at the last test failure. This is a problem with floating-point accuracy. My suggestion would be to slightly relax `region.__contains__` by adding `numpy.isclose` tests for `pmin` and `pmax`. This would mean going from `other.pmin >= self.pmin` to `other.pmin > or "approx equal" self.pmin`. Have you got any other ideas?

**Comment by marijanbeg** [comment]
@lang-m, I agree. I am just not sure what is the right way of avoiding hard-coded tolerances. Maybe we could define some tolerances in `utils` and use them throughout `df`.

**Comment by lang-m** [comment]
@marijanbeg Do you think sticking to the default settings for `rtol` and `atol` of `np.isclose` wouldn't be sufficient?

**Comment by lang-m** [comment]
The new `_as_array` implementation depending on `df.Field` cannot be in `dfu` because we run into issues with circular imports. To make it consistent I decided to move everything to the `field` module as this also is the only place where we need theses functions.

**Comment by marijanbeg** [comment]
Nice, I like that.

**Comment by marijanbeg** [comment]
@lang-m, default value for `rtol` is fine. Only `atol` could be an issue because the values we have in the field can be in a wide range. Probably setting it to zero is fine.

**Comment by lang-m** [comment]
@marijanbeg Setting `atol` to 0 does not work when one of the points is at `(0, 0, 0)` (or probably also just 0 in one direction, I didn't check that). I've now introduced a different way of calculating `rtol` and `atol`. Could you please review [d789e1c](https://github.com/ubermag/discretisedfield/pull/135/commits/d789e1c833006523bb32968b62e182915ddd834f).

**Comment by marijanbeg** [comment]
It does exactly what it says on the tin.

**Pull request merged**